### ISSUE TITLE
Need a way to update dapp data on block replacement without rescanning chain

### ISSUE THREAD

**Issue by jferas** [SushiChain/SushiChain#340]
## Checklist
- [ ] You've tried latest version of SushiChain
- [ ] You've confirmed there is not a same report

## Environment

- Using docker: (yes or no)
- Crystal version: x.x.x
- Where do you connect to? (http://testnet.sushichain.io:3000, http://localhost:300 or so on)

## Command Lines and results

```bash
> [your command]
[the result]
```

**Comment by jferas** [comment]
When a block is replaced in the chain, either via sync or single block replacement after a peer broadcast... currently the only way to re-sync the dapp/buildin data structures (including, but not limited to indices, scars, token, transaction, and utx0) is to completely clear the structures and re-create them by scanning the entire chain (this is done in dapp_clear_record).

This approach isn't scalable when the chain is even slightly long.. it will be a huge performance impact.  Work needs to be so that the dapp/buildin data structures are all reversible for a given block.  I've implemented an "unrecord" method in the "indices.cr" dapp/buildin code, since that data structure maps block IDs to transaction IDs, reversing individual blocks transactions is straightforward.

After looking at the other dapp/buildin data structures, I realized more work will need to be done there to make their storage single-block reversible.
